Sources/TuistCore/Graph/ModelExtensions/Target+Core.swift

```diff
@@ -18,8 +18,31 @@ public enum TargetError: FatalError, Equatable {
 }
 
 extension Target {
-    /// Returns the product name including the extension.
+    /// Returns the product name including the extension
+    /// if the PRODUCT_NAME build setting of the target is set and contains a static value that's consistent
+    /// throughout all the configurations, it uses that value, otherwise it defaults to the target's default.
     public var productNameWithExtension: String {
+        var settingsProductNames: Set<String> = Set()
+
+        if let value = settings?.base["PRODUCT_NAME"], case let SettingValue.string(baseProductName) = value {
+            settingsProductNames.insert(baseProductName)
+        }
+        settings?.configurations.values.forEach { configuration in
+            if let value = configuration?.settings["PRODUCT_NAME"],
+               case let SettingValue.string(configurationProductName) = value
+            {
+                settingsProductNames.insert(configurationProductName)
+            }
+        }
+
+        let productName: String
+
+        if settingsProductNames.count == 1, !settingsProductNames.first!.contains("$") {
+            productName = settingsProductNames.first!
+        } else {
+            productName = self.productName
+        }
```

Comment:
Should we lint that target always has just one product name and it doesn't use a variable? Otherwise, this is undefined behavior.

Reply:
Do you mean validating the emptiness of the `self.productName` string?

Reply:
I mean validating there's only one product name (it doesn't differ between configurations) and that it doesn't use a variable with `$` (which we already guard against)

Reply:
I skimmed through the usages, and most seem to assume that, so I think it makes sense to validate that. I'm not sure how strict we should be about that. We can start with just a warning to get a sense of how many users are in this state, and increase the strictness later on. Thoughts?

Reply:
Yeah, that sounds sensible 👍